runtime/meta/ffi.cpp

```diff
@@ -61,50 +64,86 @@ extern "C" {
-      return &ffi_type_sint64;
-    } else if (symbol == tag_type_float()) {
-      return &ffi_type_float;
-    } else if (symbol == tag_type_double()) {
-      return &ffi_type_double;
-    } else if (symbol == tag_type_uchar()) {
-      return &ffi_type_uchar;
-    } else if (symbol == tag_type_schar()) {
-      return &ffi_type_schar;
-    } else if (symbol == tag_type_ushort()) {
-      return &ffi_type_ushort;
-    } else if (symbol == tag_type_sshort()) {
-      return &ffi_type_sshort;
-    } else if (symbol == tag_type_uint()) {
-      return &ffi_type_uint;
-    } else if (symbol == tag_type_sint()) {
-      return &ffi_type_sint;
-    } else if (symbol == tag_type_ulong()) {
-      return &ffi_type_ulong;
-    } else if (symbol == tag_type_slong()) {
-      return &ffi_type_slong;
-    } else if (symbol == tag_type_longdouble()) {
-      return &ffi_type_longdouble;
-    } else {
-      throw std::invalid_argument("Arg is not a supported type");
+  static ffi_type * getTypeFromBlock(block * elem) {
+    if ((uint64_t) elem & 1) {
+      uint64_t symbol = (uint64_t) elem;
+
+      if (symbol == tag_type_void()) {
+        return &ffi_type_void;
+      } else if (symbol == tag_type_uint8()) {
+        return &ffi_type_uint8;
+      } else if (symbol == tag_type_sint8()) {
+        return &ffi_type_sint8;
+      } else if (symbol == tag_type_uint16()) {
+        return &ffi_type_uint16;
+      } else if (symbol == tag_type_sint16()) {
+        return &ffi_type_sint16;
+      } else if (symbol == tag_type_uint32()) {
+        return &ffi_type_uint32;
+      } else if (symbol == tag_type_sint32()) {
+        return &ffi_type_sint32;
+      } else if (symbol == tag_type_uint64()) {
+        return &ffi_type_uint64;
+      } else if (symbol == tag_type_sint64()) {
+        return &ffi_type_sint64;
+      } else if (symbol == tag_type_float()) {
+        return &ffi_type_float;
+      } else if (symbol == tag_type_double()) {
+        return &ffi_type_double;
+      } else if (symbol == tag_type_uchar()) {
+        return &ffi_type_uchar;
+      } else if (symbol == tag_type_schar()) {
+        return &ffi_type_schar;
+      } else if (symbol == tag_type_ushort()) {
+        return &ffi_type_ushort;
+      } else if (symbol == tag_type_sshort()) {
+        return &ffi_type_sshort;
+      } else if (symbol == tag_type_uint()) {
+        return &ffi_type_uint;
+      } else if (symbol == tag_type_sint()) {
+        return &ffi_type_sint;
+      } else if (symbol == tag_type_ulong()) {
+        return &ffi_type_ulong;
+      } else if (symbol == tag_type_slong()) {
+        return &ffi_type_slong;
+      } else if (symbol == tag_type_longdouble()) {
+        return &ffi_type_longdouble;
+      }
+    } else if (elem->h.hdr == (uint64_t)getTagForSymbolName(TYPETAG(struct))){
+      struct list * elements = (struct list *) *elem->children;
+      size_t numFields = hook_LIST_size_long(elements);
+      std::vector<ffi_type *> structElements;
```

Comment:
I think we can avoid the double allocation here and just write directly to structTypes->elements.